pkg/kubevirt/client.go

```diff
@@ -270,64 +291,85 @@ func (c *client) CreateVolumeSnapshot(ctx context.Context, namespace, name, clai
 	if err != nil {
 		klog.V(2).Infof("Error getting storage class name for claim %s in namespace %s: %v", claimName, namespace, err)
-		return "", fmt.Errorf("unable to determine snapshot class name for infra source volume")
-	}
-	if storageClassName == "" {
-		return "", fmt.Errorf("unable to determine storage class name for snapshot creation")
-	}
-	allowed, err := c.isStorageClassAllowed(ctx, storageClassName)
-	if err != nil {
-		return "", err
-	}
-	if !allowed {
-		return "", fmt.Errorf("storage class %s is not allowed for snapshot creation", storageClassName)
+		return "", fmt.Errorf("unable to determine volume snapshot class name for infra source volume")
+	}
+	if storageClassName == "" && !c.storageClassEnforcement.AllowDefault {
+		return "", fmt.Errorf("unable to determine volume snapshot class name for snapshot creation, and default not allowed")
+	} else if storageClassName != "" && !(util.Contains(c.storageClassEnforcement.AllowList, storageClassName) || c.storageClassEnforcement.AllowAll) {
+		return "", fmt.Errorf("unable to determine volume snapshot class name for snapshot creation, no valid snapshot classes found")
+	}
+	snapshotClassNames := c.getInfraSnapshotClassesFromStorageClassName(storageClassName)
+	if util.Contains(snapshotClassNames, snapshotClassName) {
+		return snapshotClassName, nil
+	}
+	if !(c.storageClassEnforcement.AllowAll || c.storageClassEnforcement.AllowDefault) {
+		tenantSnapshotClasses := c.getTenantSnapshotClassesFromInfraStorageClassName(storageClassName)
+		if len(tenantSnapshotClasses) > 0 {
+			if snapshotClassName == "" {
+				return "", fmt.Errorf("unable to determine volume snapshot class name for snapshot creation, valid snapshot classes are %v", tenantSnapshotClasses)
+			} else {
+				return "", fmt.Errorf("volume snapshot class %s is not compatible with PVC with storage class %s, valid snapshot classes for this pvc are %v", snapshotClassName, storageClassName, tenantSnapshotClasses)
+			}
+		} else {
+			return "", fmt.Errorf("unable to determine volume snapshot class name for snapshot creation, no valid snapshot classes found")
+		}
 	}
-	snapshotClass, err := c.getSnapshotClassFromStorageClass(ctx, storageClassName, snapshotClassName)
-	if err != nil {
-		return "", err
+	return "", nil
+}
+
+func (c *client) getInfraSnapshotClassesFromStorageClassName(storageClassName string) []string {
+	for _, storageSnapshotMapping := range c.infraTenantStorageSnapshotMapping {
+		infraStorageClasses := []string{}
+		infraSnapshotClasses := []string{}
+		for _, storageClasses := range storageSnapshotMapping.StorageClasses {
+			infraStorageClasses = append(infraStorageClasses, storageClasses.Infra)
+		}
+		for _, snapshotClasses := range storageSnapshotMapping.VolumeSnapshotClasses {
+			infraSnapshotClasses = append(infraSnapshotClasses, snapshotClasses.Infra)
+		}
+		if util.Contains(infraStorageClasses, storageClassName) {
+			return infraSnapshotClasses
+		}
```

Comment:
i'd find this pattern easier to understand and slightly more efficient. 

```
	for _, storageSnapshotMapping := range c.infraTenantStorageSnapshotMapping {
		infraSnapshotClasses := []string{}
                found := false
		for _, storageClasses := range storageSnapshotMapping.StorageClasses {
                        if storageClasses.Infra == storageClassName {
                             found = true
                             break
                        }
		}
                if !found {
                    continue
                }

		for _, snapshotClasses := range storageSnapshotMapping.VolumeSnapshotClasses {
			infraSnapshotClasses = append(infraSnapshotClasses, snapshotClasses.Infra)
		}
		return infraSnapshotClasses
      }
```

Reply:
Done